source/s3_auto_ranged_put.c

```diff
@@ -556,29 +556,41 @@ static bool s_s3_auto_ranged_put_update(
 
             if (should_create_next_part_request) {
 
-                /* Allocate a request for another part. */
-                request = aws_s3_request_new(
-                    meta_request,
-                    AWS_S3_AUTO_RANGED_PUT_REQUEST_TAG_PART,
-                    0,
-                    AWS_S3_REQUEST_FLAG_RECORD_RESPONSE_HEADERS);
+                struct aws_s3_buffer_pool_ticket *ticket 
+                    = aws_s3_buffer_pool_reserve(meta_request->client->buffer_pool, meta_request->part_size);
 
-                request->part_number = auto_ranged_put->threaded_update_data.next_part_number;
+                if (ticket) {
+                    /* Allocate a request for another part. */
+                    request = aws_s3_request_new(
+                        meta_request,
+                        AWS_S3_AUTO_RANGED_PUT_REQUEST_TAG_PART,
+                        0,
+                        AWS_S3_REQUEST_FLAG_RECORD_RESPONSE_HEADERS | AWS_S3_REQUEST_FLAG_PART_SIZE_REQUEST_BODY);
 
-                /* If request was previously uploaded, we prepare it to ensure checksums still match,
-                 * but ultimately it gets marked no-op and we don't send it */
-                request->was_previously_uploaded = request_previously_uploaded;
+                    request->part_number = auto_ranged_put->threaded_update_data.next_part_number;
 
-                ++auto_ranged_put->threaded_update_data.next_part_number;
-                ++auto_ranged_put->synced_data.num_parts_started;
-                ++auto_ranged_put->synced_data.num_parts_pending_read;
+                    /* If request was previously uploaded, we prepare it to ensure checksums still match,
+                    * but ultimately it gets marked no-op and we don't send it */
+                    request->was_previously_uploaded = request_previously_uploaded;
 
-                AWS_LOGF_DEBUG(
-                    AWS_LS_S3_META_REQUEST,
-                    "id=%p: Returning request %p for part %d",
-                    (void *)meta_request,
-                    (void *)request,
-                    request->part_number);
+                    request->ticket = ticket;
+
+                    ++auto_ranged_put->threaded_update_data.next_part_number;
+                    ++auto_ranged_put->synced_data.num_parts_started;
+                    ++auto_ranged_put->synced_data.num_parts_pending_read;
+
+                    AWS_LOGF_DEBUG(
+                        AWS_LS_S3_META_REQUEST,
+                        "id=%p: Returning request %p for part %d",
+                        (void *)meta_request,
+                        (void *)request,
+                        request->part_number);
+                } else {
+                    AWS_LOGF_DEBUG(
+                        AWS_LS_S3_META_REQUEST,
+                        "id=%p: Failed to allocate due to lack of memory",
```

Comment:
mention "exceeding memory limit". "lack of memory" is going to make people think malloc failed